common/flatpak-dir.c

```diff
@@ -10447,17 +10612,24 @@ flatpak_dir_find_remote_related (FlatpakDir *self,
 
               extension_ref = g_build_filename ("runtime", extension, parts[2], branch, NULL);
 
-              if (flatpak_summary_lookup_ref (summary, extension_collection_id, extension_ref, &checksum, NULL))
+              if ((summary == NULL && flatpak_cache_lookup_ref (xa_cache, extension_ref)) ||
+                  (summary != NULL && flatpak_summary_lookup_ref (summary, extension_collection_id, extension_ref, &checksum, NULL)))
                 {
                   add_related (self, related, extension, extension_collection_id, extension_ref, checksum, no_autodownload, download_if, autodelete, locale_subset);
                 }
               else if (subdirectories)
                 {
-                  g_auto(GStrv) refs = flatpak_summary_match_subrefs (summary, extension_collection_id, extension_ref);
+                  g_auto(GStrv) refs;
+                  if (summary != NULL)
+                    refs = flatpak_summary_match_subrefs (summary, extension_collection_id, extension_ref);
+                  else
+                    refs = flatpak_cache_match_subrefs (xa_cache, extension_ref);
```

Comment:
Similarly, it would be good to merge `flatpak_summary_match_subrefs()` and `flatpak_cache_match_subrefs()`.

Reply:
I am worried that someone will write some code in flatpak in future which just calls `flatpak_summary_match_subrefs()`, and forgets to handle the `flatpak_cache_match_subrefs()` case too, leading to a regression in P2P support. Similarly with the other functions here. When this is submitted upstream, I strongly suggest you find a way to factor them together (or factor this out into a higher level function `flatpak_repo_metadata_match_subrefs()` which grabs the `summary` or `ostree-metadata` and matches as appropriate), or there will be regressions in future.

Reply:
That sounds like a valid concern; thanks for explaining. I'll think about what can be done. Cosimo said this won't be ready in time for the demo anyway, so maybe we should hold off on merging it until we get something upstream.